sw/device/mask_rom/boot.md

```diff
@@ -1,71 +1,201 @@
+    peripheral_lockdown(current_rom_ext_manifest);
+
+    // PMP Region for ROM_EXT (2.c.v)
+    // **Open Q:** Integration with Secure Boot Hardware
+    // **Open Q:** Do we need to lock down Mask ROM at final jump? How sensitive
+    //             is the mask rom itself? We may end up with a jump routine
+    //             just past a page boundary, as SRAM is non-executable.
+    //             Alternatively, we may only lock some regions of the mask rom.
+    pmp_unlock_rom_ext();
+
+    // Transfer Execution to ROM_EXT (2.c.vi)
+    // **Open Q:** Integration with Secure Boot Hardware
+    // **Open Q:** Accumulated measurements to tie manifest to hardware config.
+    if (!final_jump_to_rom_ext(current_rom_ext_manifest)) {
+      // Boot Failure (no policy check)
+      // - Because we may have locked some write-enable bits that any following
+      //   manifest cannot change if it needs to, we have to just reboot here.
+      boot_failed(boot_policy, current_rom_ext_manifest);
+    }
   }
 
-  // Verify Signature (2.c.iii)
-  valid, signature = verify_rom_ext_signature(rom_ext_manifest);
-  if (!valid) {
-    // Boot Failure (check policy)
-  }
+  // Boot failed for all ROM_EXTs allowed by boot policy
+  boot_failed(boot_policy);
+}
+```
 
-  // System State Measurements (2.c.iv)
-  measurements = perform_system_state_measurements();
-  if (!boot_allowed(measurements)) {
-    // Lifecycle failure (no policy check)
-  }
+3. Execution Enters ROM_EXT stage.
 
-  // CreatorRootKey (2.c.iv)
-  root_key_identity = derive_creator_root_key(measurements);
-  load_root_key(root_key_identity);
+   Not covered by this document. Refer to Secure Boot document instead.
 
-  // PMP Region for ROM_EXT (2.c.v)
-  //   Open Q: Do we need to lock down Boot ROM at final jump?
-  pmp_unlock_rom_ext();
+## Cleaning Device State
 
-  // Lock down Peripherals based on descriptors in rom_ext
-  peripheral_lockdown(rom_ext_manifest);
+Part of this process is done before we can execute any C code. In particular, we
+have to zero all registers and all of the main RAM before we setup the CRT
+(which allows us to execute any C code). This has to be done in assembly, as we
+cannot execute C without setting up the CRT.
 
-  // Transfer Execution to ROM_EXT (2.c.vi)
-  if (!final_jump_to_rom_ext(rom_ext_manifest, signature)) {
-    // Boot Failure (check_policy)
-  }
+We want to do as little as possible in the hand-written assembly, because it is
+so hard to verify and to test. Additionally, we want to use the DIFs (which are
+written in C) because we have verified them, rather than duplicating their
+functionality in assembly, if at all possible. So this means we wait until we
+are executing C to do the final parts of chip reset, especially parts that may
+depend on reading or writing any device state.
+
+```c
+// Will actually be implemented in assembly
+clean_device_state_part_1() {
+  // - Zero all ISA Registers
+  // - Enable SRAM Scrambling
```

Comment:
i've sent you the doc on the various retention policies. Have a look and let me know what you think.
Shouldnt the 'clean normal ram' be here as well? Because that's the one ram that should get wiped no matter the circumstance.  Some rough math assuming a tight write loop would take 0.5~1mS to wipe the 128KB ram.

Do have a look at the doc though, it's worth discussing what we feel comfortable with, as in...do we HAVE to wipe, or are we okay just reloading the key.

Reply:
@cdgori 
it just occurs to me, it would probably be very bad if we cleared the memory 'after' we enabled scrambling, especially if the value we write is known (for example all 0's and all 1's). It's a different story if the value we write is not known. 

The memory scrambling is using CTR, which means if we blast it with a known pattern, if any attack is able to read back the memory contents (even scrambled), they would now know the XOR mask for a particular address.   This could become a major weakness.  So I would recommend clearing the memory first and then arming or re-arming the scramble.

Reply:
Oh interesting point about not zeroing ram after we enable scrambling.

Are we sure the same isn't true if we write zeros to all of RAM, then enable scrambling? Presumably there's a chance there of also leaking the mask for an address if it was a known value before it was scrambled.

It almost sounds like we should just trigger SRAM scrambling, without zeroing SRAM either before or after (except for the parts we need to zero for running C).

Reply:
i think it's slightly different. Let me present my thinking, and you can tell me if it makes sense.

Let's assume the RAM is already scrambled and we just booted into ROM.

1. If we just 0 it directly, the only information we gain is the hamming delta between the scrambled values and the 0.  Since we don't know the previous xor masking, we don't anything about the original plain data. So if an attacker reads out the memory, they either get the scrambled value or 0, neither of which tells them anything about the real plain data. 

2. If we enable scramble, and then start 0-ing, we are going to write 0 XOR mask into every address.  So now if an attacker reads out the memory, they just get the mask.  And since the mask is going to be the same per memory location for the duration of that power cycle, our scrambling scheme would be broken.  Again, this is contingent on someone being able to read out the memory, which is not a guarantee.  However, comparing the two options, i think the option 1 is safer.

Reply:
Ah, ok, I see. So we can (and maybe should always) zero before we enable scrambling.